FLINK-10114: Add ORC BulkWriter support for StreamingFileSink

zenfenan opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change
This pull request adds support for ORC BulkWriter for the `StreamingFileSink`.

## Brief change log

  - Add OrcBulkWriter that uses ORC's `WriterImpl` and a slightly customised implementation of ORC's `PhysicalWriter` to write elements in the form of `VectorizerRowBatch`.
  - Add OrcBulkWriterFactory to create an instance of ORCBulkWriter

## Verifying this change

This change added tests and can be verified as follows:

  - Added an integration test to test this writer and also to verify that no Serialization related exception is thrown.
  - Added a unit test to test this new writer
  - Manually verified the change by running on a cluster with a streaming source, and writing to the filesystem

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **don't know**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **yes**
  - If yes, how is the feature documented? **JavaDocs**

zenfenan commented:
@kl0u Got it. I can do that. Where should that go in? Is it okay if I just add them in Javadoc comments itself?

kl0u commented:
@zenfenan it should be [here](https://github.com/apache/flink/blob/master/docs/dev/connectors/streamfile_sink.md) at the section about `BulkFormats`.

zenfenan commented:
Cool. I'll update the PR.

kl0u commented on flink-formats/flink-orc-compress/src/test/java/org/apache/flink/formats/orc/writers/OrcBulkWriterITCase.java:
```diff
@@ -0,0 +1,104 @@
+
+			Reader reader = OrcFile.createReader(new org.apache.hadoop.fs.Path(partFile.toURI()), readerOptions);
+
+			assertSame(reader.getCompressionKind(), CompressionKind.ZLIB);
```
Here you do not specify `ZLIB` as the compression scheme but you rely on the fact that this is the default. Could you please add it explicitly (or even better, another scheme) through the configuration? This will make the test self-contained so that anyone who reads it does not have to know what are the defaults.

zenfenan replied:
Sure. Will do.

kl0u commented on flink-formats/flink-orc-compress/src/test/java/org/apache/flink/formats/orc/writers/OrcBulkWriterTest.java:
```diff
@@ -0,0 +1,134 @@
+			RecordReader recordReader = reader.rows();
+			VectorizedRowBatch batch = reader.getSchema().createRowBatch();
+
+			assertSame(reader.getCompressionKind(), CompressionKind.ZLIB);
```
Same as before. Also the validate method seems to share a lot of code with the one in the `ITCase`. Can you deduplicate it? e.g. introduce some `Util` class in the tests.

zenfenan replied:
Yep. Makes sense.

JingsongLi commented on flink-formats/flink-orc-compress/src/main/java/org/apache/flink/formats/orc/vectorizer/Vectorizer.java:
```diff
@@ -0,0 +1,46 @@
+	 * @return The VectorizedRowBatch containing the ColumnVectors of the input element
+	 * @throws IOException if there is an error while transforming the input.
+	 */
+	VectorizedRowBatch vectorize(T element) throws IOException;
```
This interface looks like one element one batch? Test class `RecordVectorizer` is this behavior. This behavior is against the ORC design.

kl0u replied:
I was also thinking about this. I went through this with the debugger and it seems that everything ends up in the same stripe (if I did not miss something), but I am not sure if compression is efficient in this case.

zenfenan replied:
@JingsongLi Earlier As @kl0u had mentioned, regardless of being a single element VectorizedRowBatch, they end up in the same stripe. TBH, I tested two approaches initially when I working on this: 
1. Vectorizer interface has two methods `addElements(T element)` which just buffers the element in a list and another method `vectorize()` which just iterates and added them into a single `VectorizedRowBatch`.
2. Vectorizer interface has one method `vectorize(T element)` which adds that element alone into a `VectorizedRowBatch` 

Upon testing, I found there was no size difference between the two approaches, so I avoided the complexity of additional buffering and went approach 2. 

Having said that, just to clarify, I will be reaching out to the ORC community to understand the difference and see if we are losing any efficiencies in this approach.

gaoyunhaii replied:
Ignore me, I did not refresh the page and do not see the above comment.

kl0u replied:
@zenfenan I think it would be really helpful if you could reach out to the ORC community to help us understand the tradeoffs in this. Thanks for taking the initiative!

JingsongLi commented on flink-formats/flink-orc-compress/src/test/java/org/apache/flink/formats/orc/writers/OrcBulkWriterTest.java:
```diff
@@ -0,0 +1,134 @@
+		for (File partFile : partFiles) {
+			assertTrue(partFile.length() > 0);
+
+			OrcFile.ReaderOptions readerOptions = OrcFile.readerOptions(new Configuration());
```
Maybe you can implement a `TestOrcInputFormat` for verify.

JingsongLi commented on flink-formats/flink-orc-compress/src/main/java/org/apache/flink/formats/orc/writers/PhysicalWriterImpl.java:
```diff
@@ -0,0 +1,368 @@
+import static org.apache.orc.impl.WriterImpl.getEstimatedBufferSize;
+
+/**
+ * A slightly customised clone of {@link org.apache.orc.impl.PhysicalFsWriter}.
```
We need add a `NOTE`, every ORC version upgrade, maybe need update this file to bugfix.

zenfenan replied:
Makes sense. I will update. :)

JingsongLi commented on flink-formats/flink-orc-compress/src/main/java/org/apache/flink/formats/orc/writers/OrcBulkWriter.java:
```diff
@@ -0,0 +1,61 @@
+
+	@Override
+	public void flush() throws IOException {
+		// We can't do anything here
```
We should do something.
ORC will keep stripe in memory, we need flush them to Flink output stream.

JingsongLi replied:
You can ignore me. Parquet do this too. Seems this method is invalid in BulkWriter now.

zenfenan replied:
Yep, that's correct. It is because of the implementation of ORCWriter which actually starts flushing the stripe once the close gets triggered. The underlying stream actually doesn't hold anything until the `flushStripe()` gets called from the ORC Writer's `close()` method so calling `flush()` is meaningless.

zenfenan commented:
@kl0u and @JingsongLi Thank you so much for reviewing the work. Regarding the the way `VectorizedRowBatch` gets handled, I will be reaching out to the ORC community with and try to understand the best practice.

I will work on the other review comments and get the changes added by this weekend. Pretty tight with work stuff at the moment. :)

gaoyunhaii commented on flink-formats/flink-orc-compress/src/main/java/org/apache/flink/formats/orc/writers/OrcBulkWriterFactory.java:
```diff
@@ -0,0 +1,134 @@
+	private final Properties writerProperties;
+	private final Map<String, String> confMap;
+	private final TypeDescription schema;
+	private OrcFile.WriterOptions writerOptions;
```
It seems that we could create the `writerOptions` in the constructor and make this field also `final`.

kl0u commented:
@zenfenan I understand that time is of an issue these days so there is no reason to hurry. The PR can wait and there is no reason for you to get stressed about it.

zenfenan commented:
@kl0u Hey Kostas. Thanks for the suggestions. I agree with the improvements in `Vectorizer`. I actually had already made changes i local to handle the `VectorizedRowBatch` like you suggested but your point of having the schema in one place is a nice +1. I will include that. However, do you really think we need add `addUserMetadata()` to the Vectorizer?

@kl0u @JingsongLi 
I got the reply from the ORC community regarding the better way a VectorizedRowBatch can be handled. The community suggested that it actually makes little difference but I think it would be better we go with a single instance of VectorizedRowBatch and handle the lifecycle functions of BulkWriter. Thoughts?

Link: https://lists.apache.org/thread.html/r7c1eb6199c2834bb60ff6e8bb2ac7cb66d8912b86164ac7a6f3aaaf4%40%3Cuser.orc.apache.org%3E

zenfenan commented:
@kl0u On the Vectorizer improvement, I believe we cannot have the addUserMetadata() in Vectorizer class because it needs the `setWriter(Writer writer)` with a writer to be called but the `Writer` is only instantiated inside the OrcBulkWriterFactory so I am suggesting that we update `OrcBulkWriterFactory` to take an optional method, say `withUserMetadata(List<OrcProto.UserMetadataItem> userMetadata)` and pass this to the OrcBulkWriter. In OrcBulkWriter's flush(), we can do something like:

```
@Override
public void flush() {
    if (this.userMetadata != null) {
        this.writer.appendUserMetadata(this.userMetadata);
    }
}
```

Thoughts?

zenfenan commented:
@kl0u @JingsongLi

I have updated the PR with the changes. Appreciate if you could take a look. Thanks :)

kl0u commented:
Thanks a lot for following up on this @zenfenan ! 

The changes that I consider the most important are the moving of the writer to the `flink-orc` module, the changes that decouple the lifecycle of the batch of that of the scope of the `addElement()` and the removal of the requirement to provide the schema twice.

The reason I proposed to add the `addUserMetadata ()` is that it seems that it is something that is meant to be exposed to the user (from the javadoc). So if users need it, we have to expose it to them. In addition, it does not seem to be much of a hustle to add it. Is there any particular reason why you would not like to add it?

zenfenan commented:
@kl0u Not at all. In fact, I have added `appendUserMetadata()` already in the latest commit which I pushed yesterday. As I have mentioned in yesterday's  comment, I don't think it can be added to the `Vectorizer` since the Writer wouldn't be created just yet. However, we can have it in the OrcBulkWriterFactory so I added that method there. 

Yesterday's commit also introduces creating the `VectorizedRowBatch` & handling its lifecycle within the `OrcBulkWriter` lifecycle and also addresses the requirement to provide the schema twice. Would appreciate if you could take a look at the changes.

And yes, I think we can move it `flink-orc` module. Some changes may have to be done since I believe the ORC dependency version between this new module and `flink-orc` is different. I'll do them.

kl0u commented:
Thanks for the work @zenfenan ! 
Are you planning to move the writer to the `flink-orc` module as discussed in the mailing list? It would be easier for me to have a look after you have integrated all the changes.

zenfenan commented:
Yes, @kl0u 

I am moving the codebase to `flink-orc` but before that I think we need to bump ORC dependency version to 1.5.6. Reason why I'm stressing on 1.5.6 is: 

1. It is the version I have used in this PR. Haha.. Anyway, if we still want to keep the ORC version as is (i.e. 1.4.3), there will be some code change in the `PhysicalFsWriterImpl`.
2. 1.4.3 which is used in the project is little outdated so it makes sense to use this as an opportunity to upgrade to an even newer and stable version.
3. 1.6.x is not being used anywhere majorly, AFAIK. Even Hive 3.x is using ORC 1.5.6 as the dependency.

Having said that, I have created a [Jira](https://issues.apache.org/jira/browse/FLINK-17142) to bump the ORC dependency version to 1.5.6. I did the change locally and tested the flink build with and everything is good so I have just raised a PR #11759. I suggest we merge that first and then this PR.

What do you think?

kl0u commented:
I do not have a strong opinion on the order. 

I suggest that @JingsongLi has a look at the PR #11759 as he can tell if this affects anything on the TableAPI or Hive side of things :) and if reviewing takes too long, then we can consider merging this first.

kl0u commented:
@zenfenan I have not check your changes yet but I had something like this https://github.com/kl0u/flink/tree/FLINK-10114 in mind for the `addUserMetadata()`. I will check your branch later to see, but feel free to check also my changes and we can chat :)

zenfenan commented:
@kl0u I checked your branch and I think it looks exactly like what you had commented earlier. My changes pretty much are the same except that `addUserMetadata()` resides in the `OrcBulkWriterFactory`. I think you will understand the rationale behind when you check my changes.

zenfenan commented:
@kl0u I have moved the implementation to `flink-orc` module. I think it should be easier to take a complete look now. :)

kl0u commented on flink-formats/flink-orc/src/main/java/org/apache/flink/orc/vector/Vectorizer.java:
```diff
@@ -0,0 +1,93 @@
+@PublicEvolving
+public abstract class Vectorizer<T> implements Serializable {
+
+	protected transient VectorizedRowBatch rowBatch;
```
I think that this should not be a field in this class, but (potentially) part of the `OrcBulkWriter`.  It is the writer that will handle the lifecycle of the batch in any case.

kl0u commented on flink-formats/flink-orc/src/main/java/org/apache/flink/orc/writer/OrcBulkWriter.java:
```diff
@@ -0,0 +1,70 @@
+
+	private final Writer writer;
+	private final Vectorizer<T> vectorizer;
+
```
The `VectorizedRowBatch` can be a member of this class, and we can initialize it in the constructor, `reset()` it in `finish()` and pass it to the `vectorize()` in the `addElement`.

zenfenan replied:
Yep. I'll update this piece.

kl0u commented on flink-formats/flink-orc/src/main/java/org/apache/flink/orc/writer/OrcBulkWriter.java:
```diff
@@ -0,0 +1,70 @@
+	private final Writer writer;
+	private final Vectorizer<T> vectorizer;
+
+	private transient List<OrcProto.UserMetadataItem> userMetadata;
```
I do not think this should be here. I was referring to the method 

```
/**
   * Add a row batch to the ORC file.
   * @param batch the rows to add
   */
  void addRowBatch(VectorizedRowBatch batch) throws IOException;
```

of the `Writer` and _not_ the `appendUserMetadata()`. The former seems to be aimed to be exposed to the user, while the latter does not (from the Javadoc).

In addition, this method seems to be intended to be called at any time, and not at the instantiation of the `OrcBulkWriter`. The user can add metadata at any point, they will be appended to a list, and the list will be appended to the output file when it is finally written. This is why in my branch I just set the `Writer` to the `vectorizer` and let the user use it whenever he/she wants. I still believe that that is the correct way, correct me if I am wrong.

zenfenan replied:
I'm assuming this is a typo and you meant `addUserMetadata` like the earlier comments and not `addRowBatch`.

kl0u commented on flink-formats/flink-orc/src/main/java/org/apache/flink/orc/writer/OrcBulkWriterFactory.java:
```diff
@@ -0,0 +1,139 @@
+	private final Map<String, String> confMap;
+
+	private OrcFile.WriterOptions writerOptions;
+	private List<OrcProto.UserMetadataItem> userMetadata;
```
If we agree on my previous comment, then this should go away.

zenfenan commented on flink-formats/flink-orc/src/main/java/org/apache/flink/orc/vector/Vectorizer.java:
```diff
@@ -0,0 +1,93 @@
+	 * @param element The input element
+	 * @throws IOException if there is an error while transforming the input.
+	 */
+	public abstract void vectorize(T element) throws IOException;
```
Makes sense.

zenfenan commented:
@kl0u I agree with the point that the VectorizedRowBatch shouldn't be exposed to the implementor. I am changing `vectorize()` method to take VectorizedRowBatch as an argument along with `T element`.

Regarding the better placement of `addUserMetadata(...)` method. My point was coming from the fact that when a user instantiates `Vectorizer`, the user cannot straightaway call `vectorizer.addUserMetadata(...)` as it will cause an NPE. Correct?

zenfenan commented on flink-formats/flink-orc/src/main/java/org/apache/flink/orc/vector/Vectorizer.java:
```diff
@@ -0,0 +1,93 @@
+		this.schema = TypeDescription.fromString(schema);
+		this.rowBatch = this.schema.createRowBatch();
+	}
+
```
Yes. When we update the vectorize() method with a new argument VectorizedRowBatch, this will go away.

kl0u commented:
In my mind, the user is going to only be able to call the `addUserMetadata()` from within the `vectorize()`, nowhere else.  So in this case, the `writer` will have already be initialized, right?

zenfenan commented:
Understood. I assume this suggestion comes from the belief that the users may update or add new metadata on the fly based on some external factor (either based on the input element or fetching from external service). Correct?

kl0u commented:
Yes, this is my understanding on how people would use such a method.

zenfenan commented:
@kl0u I am really sorry I couldn't focus on this one in the last few days. 

BTW, the reason why I originally had a different metadata method was because regardless of what the user calls, be it `addUserMetadata(key, value)` or `appendUserMetadata(metadataItems)`, the supplied metadata goes to a Map which is then called during the write process.

Anyways, I have changed them as per your suggestion. We now have a method in `Vectorizer` and I have also updated the docs explaining the same. Can you please review it now?

kl0u commented:
Yes, @zenfenan , I will do it either today or at most tomorrow. Thanks a lot for the work :)

zenfenan commented:
@kl0u Sounds good :)
